walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,58 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.stream.Collectors;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.stream.Stream;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> possibleTimes = new ArrayList();
+    Stream.concat(events.stream()
+    	.filter(event -> areAnyAttendeeBusy(event, request))
+    	.map(Event::getWhen), Stream.of(TimeRange.fromStartDuration(TimeRange.END_OF_DAY+1, 0)))
+    .sorted(TimeRange.ORDER_BY_START)
+    .reduce(TimeRange.fromStartDuration(0, 0), (previous, current) -> 
+      mergeTimes(previous, current, request, possibleTimes)
+    );
+
+    return possibleTimes;
+  }
+
+  public Boolean areAnyAttendeeBusy(Event event, MeetingRequest request) {
+    return !event.getAttendees()
+      .stream()
+      .filter(attendee->request.getAttendees().contains(attendee))
+      .collect(Collectors.toList())
+      .isEmpty();
+  }
+
+  //Merges two invalid times based on their overlap
```

Comment:
Awesome approach to the problem!

Great job breaking it up into different cases and adding comments where appropriate.

Reductions can be extremely complicated but you made it simple to understand.